rosbag2_performance/rosbag2_performance_benchmarking/src/benchmark_publishers.cpp

```diff
@@ -109,20 +117,34 @@ class BenchmarkPublishers : public rclcpp::Node
     producer->period = std::chrono::milliseconds(
       producer_config.frequency ? 1000 / producer_config.frequency : 1);
 
+    thread_pool_.queue(
+      [this, initial_time, producer] {
+        producer_job(initial_time, producer);
+      });
+
     return producer;
   }
 
-  void producer_job(std::shared_ptr<BenchmarkProducer> producer)
+  void producer_job(
+    std::chrono::time_point<std::chrono::high_resolution_clock> when,
+    std::shared_ptr<BenchmarkProducer> producer)
   {
-    for (auto i = 0u; i < producer->max_messages; ++i) {
-      if (!rclcpp::ok()) {
-        break;
-      }
-      std::this_thread::sleep_for(producer->period);
-      producer->produce();
+    std::this_thread::sleep_until(when);
+    producer->produce();
```

Comment:
I think we need to add check for 
```
      if (!rclcpp::ok()) {
        producer->promise_finished.set_value();
        return;
      }
```
Before calling `producer->produce();` or at least before `std::this_thread::sleep_until(when);`

Reply:
done